expression/builtin_arithmetic_vec.go

```diff
@@ -429,13 +429,101 @@ func (b *builtinArithmeticDivideRealSig) vecEvalReal(input *chunk.Chunk, result
 }
 
 func (b *builtinArithmeticIntDivideIntSig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinArithmeticIntDivideIntSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	lh, err := b.bufAllocator.get(types.ETInt, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(lh)
+	if err := b.args[0].VecEvalInt(b.ctx, input, lh); err != nil {
+		return err
+	}
+	if err := b.args[1].VecEvalInt(b.ctx, input, result); err != nil {
+		return err
+	}
+	result.MergeNulls(lh)
+	rh := result
+	dividend := lh.Int64s()
+	divisor := rh.Int64s()
+	for i := 0; i < len(divisor); i++ {
+		if !rh.IsNull(i) && divisor[i] == 0 {
+			if err := handleDivisionByZeroError(b.ctx); err != nil {
+				return err
+			}
+			result.SetNull(i, true)
+		}
+	}
+	resulti64s := result.Int64s()
+	isLHSUnsigned := mysql.HasUnsignedFlag(b.args[0].GetType().Flag)
+	isRHSUnsigned := mysql.HasUnsignedFlag(b.args[1].GetType().Flag)
+	switch {
+	case isLHSUnsigned && isRHSUnsigned:
+		err = b.divideUU(dividend, divisor, resulti64s)
+	case isLHSUnsigned && !isRHSUnsigned:
+		err = b.divideUI(dividend, divisor, resulti64s)
+	case !isLHSUnsigned && isRHSUnsigned:
+		err = b.divideIU(dividend, divisor, resulti64s)
+	case !isLHSUnsigned && !isRHSUnsigned:
+		err = b.divideII(dividend, divisor, resulti64s)
+	}
+	return err
+}
+func (b *builtinArithmeticIntDivideIntSig) divideUU(dividend, divisor, resulti64s []int64) error {
+	for i := 0; i < len(dividend); i++ {
+		if divisor[i] == 0 {
+			continue
+		}
+		lh, rh := uint64(dividend[i]), uint64(divisor[i])
+		resulti64s[i] = int64(lh / rh)
+	}
+	return nil
+}
+func (b *builtinArithmeticIntDivideIntSig) divideUI(dividend, divisor, resulti64s []int64) error {
+	for i := 0; i < len(dividend); i++ {
+		if divisor[i] == 0 {
+			continue
+		}
+		lh, rh := uint64(dividend[i]), divisor[i]
+		val, err := types.DivUintWithInt(lh, rh)
+		if err != nil {
+			return err
+		}
+		resulti64s[i] = int64(val)
+	}
+	return nil
+}
+func (b *builtinArithmeticIntDivideIntSig) divideIU(dividend, divisor, resulti64s []int64) error {
+	for i := 0; i < len(dividend); i++ {
+		if divisor[i] == 0 {
+			continue
+		}
+		lh, rh := dividend[i], uint64(divisor[i])
+		val, err := types.DivIntWithUint(lh, rh)
+		if err != nil {
+			return err
+		}
+		resulti64s[i] = int64(val)
+	}
+	return nil
+}
+func (b *builtinArithmeticIntDivideIntSig) divideII(dividend, divisor, resulti64s []int64) error {
+	for i := 0; i < len(dividend); i++ {
+		if divisor[i] == 0 {
+			continue
+		}
+		lh, rh := dividend[i], divisor[i]
+		val, err := types.DivInt64(lh, rh)
```

Comment:
It seems that we do not need to call `types.DivInt64` here. MySQL will not raise an error for this case.

``` sql
mysql> select -1 / -9223372036854775808;
+---------------------------+
| -1 / -9223372036854775808 |
+---------------------------+
|                    0.0000 |
+---------------------------+
1 row in set (0.00 sec)

mysql> drop table t;
Query OK, 0 rows affected (0.05 sec)

mysql> create table t(a int8);
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t values(-1 / -9223372036854775808);
Query OK, 1 row affected (0.01 sec)

mysql> select * from t;
+------+
| a    |
+------+
|    0 |
+------+
1 row in set (0.00 sec)
```

Please check the case that will cause an error in `DivIntWithUint` and `DivUintWithInt` in MySQL.

Reply:
Did you check whether we need `DivIntWithUint` and `DivUintWithInt`?